Index: ft_http_server.c
===================================================================
RCS file: /cvsroot/gift/OpenFT/src/ft_http_server.c,v
retrieving revision 1.59
diff -u -r1.59 ft_http_server.c
--- ft_http_server.c	23 Dec 2003 17:56:12 -0000	1.59
+++ ft_http_server.c	14 Apr 2004 00:40:22 -0000
@@ -270,7 +286,6 @@
 	assert (decoded != NULL);
 
 	spath = file_secure_path (decoded);
-	assert (spath != NULL);
 
 	free (decoded);
 
@@ -279,7 +294,7 @@
 	 * snapshot of the [connected] nodes cache for debugging/diagnostic
 	 * purposes.  This may be replaced in the future...
 	 */
-	if (strcmp (spath, "/nodes") == 0)
+	if (spath && strcmp (spath, "/nodes") == 0)
 	{
 		if ((share = access_nodes_share (req)))
 			ret = UPLOAD_AUTH_ALLOW;
@@ -293,8 +308,8 @@
 		 * before we can request authorization from giFT.  Then, we need to
 		 * actually ask giFT for authorization.
 		 */
-		if ((share = FT->share_lookup (FT, SHARE_LOOKUP_HPATH, spath)))
+		if (spath && (share = FT->share_lookup (FT, SHARE_LOOKUP_HPATH, spath)))
 			ret = FT->upload_auth (FT, net_ip_str (c->host), share, auth_info);
 		else
 			ret = UPLOAD_AUTH_NOTSHARED;
 	}
