Index: src/meta_ogg.c
===================================================================
RCS file: /cvsroot/gift/giFT/src/meta_ogg.c,v
retrieving revision 1.1
diff -u -r1.1 meta_ogg.c
--- src/meta_ogg.c	11 Sep 2002 00:38:16 -0000	1.1
+++ src/meta_ogg.c	12 Sep 2002 18:08:21 -0000
@@ -56,9 +56,8 @@
 	freq    = vinfo->rate;
 	dur     = (int) ov_time_total (&oggfile, -1);
 
-	/* sigh, we have to use kbps */
 	if (bitrate > 0)
-		meta_set (file, "Bitrate", stringf ("%li", bitrate / 1000));
+		meta_set (file, "Bitrate", stringf ("%li", bitrate));
 
 	meta_set (file, "Frequency", stringf ("%li", freq));
 	meta_set (file, "Duration",  stringf ("%i", dur));
